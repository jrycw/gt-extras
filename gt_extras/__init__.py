
# Import objects from the module
from .themes import gt_theme_538
from .themes import gt_theme_espn

__all__ = ["gt_theme_538", "gt_theme_espn"]