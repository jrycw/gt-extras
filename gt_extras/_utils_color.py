from __future__ import annotations

from great_tables._data_color.base import (
    _get_domain_factor,
    _html_color,
    _rescale_factor,
)
from great_tables._data_color.constants import ALL_PALETTES, DEFAULT_PALETTE
from great_tables._data_color.palettes import GradientPalette
from great_tables._tbl_data import TblData

__all__ = ["_get_discrete_colors_from_palette"]


def _get_discrete_colors_from_palette(
    palette: list[str] | str | None, data: list, data_table: TblData
) -> list[str | None]:
    # If palette is not provided, use a default palette
    if palette is None:
        palette = DEFAULT_PALETTE

    # Otherwise get the palette from great_tables._data_color
    elif isinstance(palette, str):
        palette = ALL_PALETTES.get(palette, [palette])

    # Standardize values in `palette` to hexadecimal color values
    palette = _html_color(colors=palette)

    # Rescale the category column for the purpose of assigning colors to each dot
    domain_factor = _get_domain_factor(df=data_table, vals=data)
    scaled_vals = _rescale_factor(
        df=data_table,
        vals=data,
        domain=domain_factor,  # type: ignore
        palette=palette,
    )

    # Create a color scale function from the palette
    color_scale_fn = GradientPalette(colors=palette)

    # Call the color scale function on the scaled categoy values to get a list of colors
    color_vals = color_scale_fn(scaled_vals)

    return color_vals
