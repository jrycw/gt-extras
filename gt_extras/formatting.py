from __future__ import annotations
from typing import Literal

from great_tables import GT
from great_tables._tbl_data import SelectExpr

__all__ = ["fmt_pct_extra"]


def fmt_pct_extra(
    gt: GT,
    column: SelectExpr,
) -> GT:
    """
    Docstring goes here
    """
    pass