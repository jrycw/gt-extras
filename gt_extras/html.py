from __future__ import annotations
from typing import Literal

__all__ = ["gt_hyperlink", "with_tooltip"]


def gt_hyperlink(text: str, url: str, new_tab: bool = True) -> int:
    """
    Create HTML hyperlinks for use in `GT` object cells.

    The `gt_hyperlink()` function creates properly formatted HTML hyperlink elements that can be
    used within table cells.

    Parameters
    ----------
    text
        A character string that will be displayed as the clickable link text.

    url
        A character string indicating the destination URL for the hyperlink.

    new_tab
        A boolean indicating whether the link should open in a new browser tab or the current tab.

    Returns
    -------
    str
        An string containing the HTML formatted hyperlink element.
    """
    target = "_self"
    if new_tab:
        target = "_blank"

    return f'<a href="{url}" target="{target}">{text}</a>'


def with_tooltip(
    label: str,
    tooltip: str,
    text_decoration_style: Literal["underline", "dotted"] | None = "dotted",
    color: str | None = "blue",
) -> str:
    """
    Create HTML text with tooltip functionality for use in GT table cells.

    The `with_tooltip()` function creates an HTML `<abbr>` element with a tooltip that appears
    when users hover over the text. The text is styled with an underline and blue color to
    indicate it's interactive.

    Parameters
    ----------
    label
        A character string that will be displayed as the visible text.

    tooltip
        A character string that will appear as the tooltip when hovering over the label.

    Returns
    -------
    str
        An HTML string containing the formatted tooltip element.
    """

    style = "cursor: help; "

    if text_decoration_style is not None:
        style += "text-decoration: underline; "
        style += f"text-decoration-style: {text_decoration_style}; "
    else:
        style += "text-decoration: none; "

    if color is not None:
        style += f"color: {color}; "

    return f'<abbr style="{style}" title="{tooltip}">{label}</abbr>'
