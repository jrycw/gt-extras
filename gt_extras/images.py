from __future__ import annotations

from great_tables import html

__all__ = ["img_header"]


def img_header(
    label: str,
    img_url: str,
    height: float = 60,
    font_size: int = 12,
    border_color: str = "black",
    text_color: str = "black",
) -> html:
    """
    Create an HTML header with an image and a label.

    Parameters
    ----------
    label
        The text label to display below the image.

    img_url
        The URL of the image to display. This can be a local filepath or an image on the web.

    height
        The height of the image in pixels.

    font_size
        The font size of the label text.

    border_color
        The color of the border below the image.

    text_color
        The color of the label text.

    Returns
    -------
    html
        A Great Tables `html` element for the header.

    Examples
    -------
    ```{python}

    ```
    """

    img_html = f"""
    <img src="{img_url}" style="
        height:{height}px;
        border-bottom:2px solid {border_color};"
    />
    """

    label_html = f"""
    <div style="
        font-size:{font_size}px;
        color:{text_color};
        text-align:center;
        width:100%;
    ">
        {label}
    </div>
    """

    full_element = f"""
    <div style="text-align:center;">
        {img_html}
        {label_html}
    </div>
    """

    return html(full_element)
