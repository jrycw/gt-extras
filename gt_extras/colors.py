from __future__ import annotations
from typing import Literal

from great_tables import GT, style, loc
from great_tables._tbl_data import SelectExpr

from great_tables._data_color.base import _html_color

__all__ = ["gt_highlight_cols", "gt_hulk_col_numeric"]


def gt_highlight_cols(
    gt: GT,
    columns: SelectExpr = None,
    fill: str = "#80bcd8",
    alpha: int | None = None,
    font_weight: Literal["normal", "bold", "bolder", "lighter"] | int = "normal",
    font_color: str = "#000000",
) -> GT:
    # TODO: see if the color can be displayed in some cool way in the docs
    """
    Add color highlighting to one or more specific columns.

    The `gt_highlight_cols()` function takes an existing `GT` object and adds highlighting color
    to the cell background of a specific column(s).

    Parameters
    ----------
    gt
        An existing `GT` object.

    columns
        The columns to target. Can either be a single column name or a series of column names
        provided in a list. If `None`, the alignment is applied to all columns.

    fill
        A character string indicating the fill color. If nothing is provided, then `"#80bcd8"`
        (light blue) will be used as a default.

    alpha
        An integer `[0, 1]` for the alpha transparency value for the color as single value in the
        range of `0` (fully transparent) to `1` (fully opaque). If not provided the fill color will
        either be fully opaque or use alpha information from the color value if it is supplied in
        the `"#RRGGBBAA"` format.

    font_weight
        A string or number indicating the weight of the font. Can be a text-based keyword such as
        `"normal"`, `"bold"`, `"lighter"`, `"bolder"`, or, a numeric value between `1` and `1000`,
        inclusive. Note that only variable fonts may support the numeric mapping of weight.

    font_color
        A character string indicating the text color. If nothing is provided, then `"#000000"`
        (black) will be used as a default.

    Returns
    -------
    GT
        The `GT` object is returned. This is the same object that the method is called on so that
        we can facilitate method chaining.

    Examples
    --------
    ```{python}
    from great_tables import GT, md
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars[["model", "year", "hp", "trq"]].head(9)

    gt = (
        GT(gtcars_mini, rowname_col="model")
        .tab_stubhead(label=md("*Car*"))
    )

    gte.gt_highlight_cols(gt, columns="hp")
    ```
    """
    if alpha:
        fill = _html_color(colors=[fill], alpha=alpha)[0]

    res = gt.tab_style(
        style=[
            style.fill(color=fill),
            style.text(weight=font_weight, color=font_color),
            style.borders(sides=["top", "bottom"], color=fill),
        ],
        locations=loc.body(columns=columns),
    )

    return res


def gt_hulk_col_numeric(
    gt: GT,
    columns: SelectExpr = None,
    palette: str | list[str] | None = "PRGn",
    domain: list[str] | list[int] | list[float] | None = None,
    na_color: str | None = None,
    alpha: int | float | None = None,  # TODO: see note
    reverse: bool = False,
    autocolor_text: bool = True,
) -> GT:
    # TODO: alpha is incomplete
    """
    Apply a color gradient to numeric columns in a `GT` table.

    The `gt_hulk_col_numeric()` function takes an existing `GT` object and applies a color gradient
    to the background of specified numeric columns, based on their values. This is useful for
    visually emphasizing the distribution or magnitude of numeric data within a table.

    Parameters
    ----------
    gt
        An existing `GT` object.

    columns
        The columns to target. Can be a single column name or a list of column names. If `None`,
        the color gradient is applied to all columns.

    palette
        The color palette to use for the gradient. Can be a string referencing a palette name or a
        list of color hex codes. Defaults to the `"PRGn"` palette from the ColorBrewer package.

    domain
        The range of values to map to the color palette. Should be a list of two values (min and
        max). If `None`, the domain is inferred from the data.

    na_color
        The color to use for missing (`NA`) values. If `None`, a default color is used.

    alpha
        The alpha (transparency) value for the colors, as a float between `0` (fully transparent)
        and `1` (fully opaque).

    reverse
        If `True`, reverses the color palette direction.

    autocolor_text
        If `True`, automatically adjusts text color for readability against the background.

    Returns
    -------
    GT
        The modified `GT` object, allowing for method chaining.

    Examples
    --------
    ```python
    from great_tables import GT
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars[["model", "year", "hp", "trq"]].head(9)

    gt = (
        GT(gtcars_mini, rowname_col="model")
        .tab_stubhead(label="Car")
    )

    gte.gt_hulk_col_numeric(gt, columns=["hp", "trq"])
    ```
    """
    res = gt.data_color(
        columns=columns,
        palette=palette,
        domain=domain,
        na_color=na_color,
        alpha=alpha,  # TODO note alpha is not supported in data_color
        reverse=reverse,
        autocolor_text=autocolor_text,
    )

    return res
