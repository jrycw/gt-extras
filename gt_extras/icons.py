from __future__ import annotations
from typing import Literal

from faicons import icon_svg

__all__ = "fa_icon_repeat"


def fa_icon_repeat(
    name: str = "star",
    repeats: int = 1,
    fill: str | None = "black",
    fill_opacity: int | str | None = 1,
    stroke: str | None = None,
    stroke_width: str | None = None,
    stroke_opacity: int | str | None = None,
    height: str | None = None,
    width: str | None = None,
    margin_left: str | None = "auto",
    margin_right: str | None = "0.2em",
    position: str | None = "relative",
    title: str | None = None,
    a11y: Literal["deco", "sem"] | None = "deco",
) -> str:
    """
    Create repeated FontAwesome SVG icons as HTML.

    The `fa_icon_repeat()` function generates one or more FontAwesome SVG icons using the `faicons`
    package and returns them as a single HTML string.

    Parameters
    ----------
    name
        The name of the FontAwesome icon to use (e.g., `"star"`, `"thumbs-up"`).

    repeats
        The number of times to repeat the icon.

    fill
        The fill color for the icon (e.g., `"yellow"`, `"#ffcc00"`). If `None`, uses the default.

    fill_opacity
        The opacity of the fill color (`0.0` - `1.0`).

    stroke
        The stroke color for the icon outline.

    stroke_width
        The width of the icon outline.

    stroke_opacity
        The opacity of the outline (`0.0` - `1.0`).

    height
        The height of the icon.

    width
        The width of the icon.

    margin_left
        The left margin for the icon.

    margin_right
        The right margin for the icon.

    position
        The CSS position property for the icon (e.g., `"absolute"`, `"relative"`, etc).

    title
        The title (tooltip) for the icon.

    a11y
        Accessibility mode: `"deco"` for decorative, `"sem"` for semantic.

    Returns
    -------
    str
        An HTML string containing the repeated SVG icons. If `repeats = 0`, this string will be empty.

    Examples
    --------
    ```{python}
    import pandas as pd
    from great_tables import GT
    import gt_extras as gte

    df = pd.DataFrame({
        "Name": ["Alice", "Bob", "Carol"],
        "Stars": [
            gte.fa_icon_repeat(name="star", repeats=3, fill="gold", fill_opacity=0.66),
            gte.fa_icon_repeat(name="star", repeats=2, fill="gold", stroke="black", stroke_width="3em"),
            gte.fa_icon_repeat(name="star", repeats=1, fill="orange"),
        ]
    })

    GT(df)
    ```

    Note
    --------
    See `icon_svg()` in the `faicons` package for further implementation details.
    """
    if repeats < 0:
        raise ValueError("repeats must be >= 0")

    icon = icon_svg(
        name=name,
        fill=fill,
        fill_opacity=fill_opacity,
        stroke=stroke,
        stroke_width=stroke_width,
        stroke_opacity=stroke_opacity,
        height=height,
        width=width,
        margin_left=margin_left,
        margin_right=margin_right,
        position=position,
        title=title,
        a11y=a11y,
    )

    repeated_icon = "".join(str(icon) for _ in range(repeats))

    return repeated_icon
