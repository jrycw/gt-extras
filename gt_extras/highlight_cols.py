from __future__ import annotations

from great_tables import GT, style, loc
from great_tables._tbl_data import SelectExpr

from typing import Literal

import matplotlib.colors as mcolors


def highlight_cols(
    gt: GT,
    columns: SelectExpr = None,  ## Todo check that this is good
    fill: str = "#80bcd8",
    alpha: int = 1,  ## Todo this doesn't exist in Python
    font_weight: Literal["normal", "bold", "bolder", "lighter"] | int = "normal",
    font_color: str = "#000000",
) -> GT:
    """
    Add color highlighting to one or more specific columns

    The `gt_highlight_cols()` function takes an existing `GT` object and adds highlighting color
    to the cell background of a specific column(s).

    Parameters
    ----------
    gt
        An existing `GT` object

    columns
        The columns to target. Can either be a single column name or a series of column names
        provided in a list. If `None`, the alignment is applied to all columns.

    fill
        A character string indicating the fill color. If nothing is provided, then `"#80bcd8"`
        (light blue) will be used as a default.
    TODO see if the color can be displayed in some cool way

    alpha??
        TODO

    font_weight
        A string or number indicating the weight of the font. Can be a text-based keyword such as
        `"normal"`, `"bold"`, `"lighter"`, `"bolder"`, or, a numeric value between `1` and `1000`,
        inclusive. Note that only variable fonts may support the numeric mapping of weight.

    font_color
        A character string indicating the text color. If nothing is provided, then `"#000000"`
        (black) will be used as a default.

    Returns
    -------
    GT
        The `GT` object is returned. This is the same object that the method is called on so that
        we can facilitate method chaining.

    Examples
    --------

    TODO
    The example below relabels columns from the `countrypops` data to start with uppercase.

    ```{python}
    from great_tables import GT
    from great_tables.data import countrypops

    countrypops_mini = countrypops.loc[countrypops["country_name"] == "Uganda"][
        ["country_name", "year", "population"]
    ].tail(5)

    (
        GT(countrypops_mini)
        .cols_label(
            country_name="Country Name",
            year="Year",
            population="Population"
        )
    )
    ```
    """

    # Altered wrt R package - no alpha
    def _to_alpha_hex_color(color: str, alpha: int) -> str:
        """
        Function description 
        TODO Can we do it without importing mcolors?
        """
        try:
            rbg_color = mcolors.to_rgb(color)        
            rbg_color_with_alpha = rbg_color + (alpha, )
            hex_color_with_alpha = mcolors.to_hex(rbg_color_with_alpha, keep_alpha=True)
            return hex_color_with_alpha
        except ValueError:
            raise ValueError(f"Invalid color value: {color}")

    alpha = min(max(alpha, 0), 1)
    fill_with_alpha = _to_alpha_hex_color(fill, alpha=alpha)

    res = gt.tab_style(
        style=[
            style.fill(color=fill_with_alpha),
            style.text(weight=font_weight, color=font_color),
            style.borders(sides=["top", "bottom"], color=fill),
        ],
        locations=loc.body(columns=columns),
    )

    return res
