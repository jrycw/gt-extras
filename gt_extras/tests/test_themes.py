from great_tables import GT, exibble
import gt_extras
# import pytest

def test_theme_538():
    assert True

def test_theme_538_snap(snapshot):
    themed_gt = gt_extras.gt_theme_538(gt=GT(exibble, id="538_table"))
    assert snapshot == themed_gt.as_raw_html()
    # TODO: make snapshot smaller (google font import + family)