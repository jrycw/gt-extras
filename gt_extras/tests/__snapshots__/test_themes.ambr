# serializer version: 1
# name: test_theme_538_snap
  '''
  <div id="538_table" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
  <style>
  @import url('https://fonts.googleapis.com/css2?family=Cairo&display=swap');
  @import url('https://fonts.googleapis.com/css2?family=Chivo&display=swap');
  #538_table table {
            font-family: Cairo, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
            -webkit-font-smoothing: antialiased;
            -moz-osx-font-smoothing: grayscale;
          }
  
  #538_table thead, tbody, tfoot, tr, td, th { border-style: none; }
   tr { background-color: transparent; }
  #538_table p { margin: 0; padding: 0; }
   #538_table .gt_table { display: table; border-collapse: collapse; line-height: normal; margin-left: auto; margin-right: auto; color: #333333; font-size: 16px; font-weight: 400; font-style: normal; background-color: #FFFFFF; width: auto; border-top-style: none; border-top-width: 3px; border-top-color: #A8A8A8; border-right-style: none; border-right-width: 2px; border-right-color: #D3D3D3; border-bottom-style: none; border-bottom-width: 2px; border-bottom-color: #A8A8A8; border-left-style: none; border-left-width: 2px; border-left-color: #D3D3D3; }
   #538_table .gt_caption { padding-top: 4px; padding-bottom: 4px; }
   #538_table .gt_title { color: #333333; font-size: 125%; font-weight: initial; padding-top: 4px; padding-bottom: 4px; padding-left: 5px; padding-right: 5px; border-bottom-color: #FFFFFF; border-bottom-width: 0; }
   #538_table .gt_subtitle { color: #333333; font-size: 85%; font-weight: initial; padding-top: 3px; padding-bottom: 5px; padding-left: 5px; padding-right: 5px; border-top-color: #FFFFFF; border-top-width: 0; }
   #538_table .gt_heading { background-color: #FFFFFF; text-align: left; border-bottom-color: #FFFFFF; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; }
   #538_table .gt_bottom_border { border-bottom-style: none; border-bottom-width: 2px; border-bottom-color: #D3D3D3; }
   #538_table .gt_col_headings { border-top-style: none; border-top-width: 2px; border-top-color: #D3D3D3; border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: black; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; }
   #538_table .gt_col_heading { color: #333333; background-color: white; font-size: 100%; font-weight: normal; text-transform: inherit; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; vertical-align: bottom; padding-top: 5px; padding-bottom: 5px; padding-left: 5px; padding-right: 5px; overflow-x: hidden; }
   #538_table .gt_column_spanner_outer { color: #333333; background-color: white; font-size: 100%; font-weight: normal; text-transform: inherit; padding-top: 0; padding-bottom: 0; padding-left: 4px; padding-right: 4px; }
   #538_table .gt_column_spanner_outer:first-child { padding-left: 0; }
   #538_table .gt_column_spanner_outer:last-child { padding-right: 0; }
   #538_table .gt_column_spanner { border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: black; vertical-align: bottom; padding-top: 5px; padding-bottom: 5px; overflow-x: hidden; display: inline-block; width: 100%; }
   #538_table .gt_spanner_row { border-bottom-style: hidden; }
   #538_table .gt_group_heading { padding-top: 8px; padding-bottom: 8px; padding-left: 5px; padding-right: 5px; color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; text-transform: inherit; border-top-style: none; border-top-width: 2px; border-top-color: black; border-bottom-style: solid; border-bottom-width: 1px; border-bottom-color: black; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; vertical-align: middle; text-align: left; }
   #538_table .gt_empty_group_heading { padding: 0.5px; color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; border-top-style: none; border-top-width: 2px; border-top-color: black; border-bottom-style: solid; border-bottom-width: 1px; border-bottom-color: black; vertical-align: middle; }
   #538_table .gt_from_md> :first-child { margin-top: 0; }
   #538_table .gt_from_md> :last-child { margin-bottom: 0; }
   #538_table .gt_row { padding-top: 3px; padding-bottom: 3px; padding-left: 5px; padding-right: 5px; margin: 10px; border-top-style: solid; border-top-width: 1px; border-top-color: #D3D3D3; border-left-style: none; border-left-width: 1px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 1px; border-right-color: #D3D3D3; vertical-align: middle; overflow-x: hidden; }
   #538_table .gt_stub { color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; text-transform: inherit; border-right-style: solid; border-right-width: 0px; border-right-color: white; padding-left: 5px; padding-right: 5px; }
   #538_table .gt_stub_row_group { color: #333333; background-color: #FFFFFF; font-size: 100%; font-weight: initial; text-transform: inherit; border-right-style: solid; border-right-width: 2px; border-right-color: #D3D3D3; padding-left: 5px; padding-right: 5px; vertical-align: top; }
   #538_table .gt_row_group_first td { border-top-width: 2px; }
   #538_table .gt_row_group_first th { border-top-width: 2px; }
   #538_table .gt_striped { background-color: rgba(128,128,128,0.05); }
   #538_table .gt_table_body { border-top-style: solid; border-top-width: 2px; border-top-color: #D3D3D3; border-bottom-style: solid; border-bottom-width: 2px; border-bottom-color: #D3D3D3; }
   #538_table .gt_sourcenotes { color: #333333; background-color: #FFFFFF; border-bottom-style: none; border-bottom-width: 2px; border-bottom-color: #D3D3D3; border-left-style: none; border-left-width: 2px; border-left-color: #D3D3D3; border-right-style: none; border-right-width: 2px; border-right-color: #D3D3D3; }
   #538_table .gt_sourcenote { font-size: 12px; padding-top: 4px; padding-bottom: 4px; padding-left: 5px; padding-right: 5px; text-align: left; }
   #538_table .gt_left { text-align: left; }
   #538_table .gt_center { text-align: center; }
   #538_table .gt_right { text-align: right; font-variant-numeric: tabular-nums; }
   #538_table .gt_font_normal { font-weight: normal; }
   #538_table .gt_font_bold { font-weight: bold; }
   #538_table .gt_font_italic { font-style: italic; }
   #538_table .gt_super { font-size: 65%; }
   #538_table .gt_footnote_marks { font-size: 75%; vertical-align: 0.4em; position: initial; }
   #538_table .gt_asterisk { font-size: 100%; vertical-align: 0; }
   
  </style>
  <table class="gt_table" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
  <thead>
  
  <tr class="gt_col_headings">
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-num">num</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-char">char</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-fctr">fctr</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-date">date</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-time">time</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-datetime">datetime</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_right" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-currency">currency</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-row">row</th>
    <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" style="border-top: 0px solid black; font-family: Chivo;font-size: 14px;vertical-align: bottom;font-weight: 200;text-transform: uppercase;" scope="col" id="538_table-group">group</th>
  </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr>
      <td class="gt_row gt_right">0.1111</td>
      <td class="gt_row gt_left">apricot</td>
      <td class="gt_row gt_left">one</td>
      <td class="gt_row gt_right">2015-01-15</td>
      <td class="gt_row gt_right">13:35</td>
      <td class="gt_row gt_right">2018-01-01 02:22</td>
      <td class="gt_row gt_right">49.95</td>
      <td class="gt_row gt_left">row_1</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">2.222</td>
      <td class="gt_row gt_left">banana</td>
      <td class="gt_row gt_left">two</td>
      <td class="gt_row gt_right">2015-02-15</td>
      <td class="gt_row gt_right">14:40</td>
      <td class="gt_row gt_right">2018-02-02 14:33</td>
      <td class="gt_row gt_right">17.95</td>
      <td class="gt_row gt_left">row_2</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">33.33</td>
      <td class="gt_row gt_left">coconut</td>
      <td class="gt_row gt_left">three</td>
      <td class="gt_row gt_right">2015-03-15</td>
      <td class="gt_row gt_right">15:45</td>
      <td class="gt_row gt_right">2018-03-03 03:44</td>
      <td class="gt_row gt_right">1.39</td>
      <td class="gt_row gt_left">row_3</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">444.4</td>
      <td class="gt_row gt_left">durian</td>
      <td class="gt_row gt_left">four</td>
      <td class="gt_row gt_right">2015-04-15</td>
      <td class="gt_row gt_right">16:50</td>
      <td class="gt_row gt_right">2018-04-04 15:55</td>
      <td class="gt_row gt_right">65100.0</td>
      <td class="gt_row gt_left">row_4</td>
      <td class="gt_row gt_left">grp_a</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">5550.0</td>
      <td class="gt_row gt_left"><NA></td>
      <td class="gt_row gt_left">five</td>
      <td class="gt_row gt_right">2015-05-15</td>
      <td class="gt_row gt_right">17:55</td>
      <td class="gt_row gt_right">2018-05-05 04:00</td>
      <td class="gt_row gt_right">1325.81</td>
      <td class="gt_row gt_left">row_5</td>
      <td class="gt_row gt_left">grp_b</td>
    </tr>
    <tr>
      <td class="gt_row gt_right"><NA></td>
      <td class="gt_row gt_left">fig</td>
      <td class="gt_row gt_left">six</td>
      <td class="gt_row gt_right">2015-06-15</td>
      <td class="gt_row gt_right"><NA></td>
      <td class="gt_row gt_right">2018-06-06 16:11</td>
      <td class="gt_row gt_right">13.255</td>
      <td class="gt_row gt_left">row_6</td>
      <td class="gt_row gt_left">grp_b</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">777000.0</td>
      <td class="gt_row gt_left">grapefruit</td>
      <td class="gt_row gt_left">seven</td>
      <td class="gt_row gt_right"><NA></td>
      <td class="gt_row gt_right">19:10</td>
      <td class="gt_row gt_right">2018-07-07 05:22</td>
      <td class="gt_row gt_right"><NA></td>
      <td class="gt_row gt_left">row_7</td>
      <td class="gt_row gt_left">grp_b</td>
    </tr>
    <tr>
      <td class="gt_row gt_right">8880000.0</td>
      <td class="gt_row gt_left">honeydew</td>
      <td class="gt_row gt_left">eight</td>
      <td class="gt_row gt_right">2015-08-15</td>
      <td class="gt_row gt_right">20:20</td>
      <td class="gt_row gt_right"><NA></td>
      <td class="gt_row gt_right">0.44</td>
      <td class="gt_row gt_left">row_8</td>
      <td class="gt_row gt_left">grp_b</td>
    </tr>
  </tbody>
  
  
  </table>
  
  </div>
          
  '''
# ---
